cri/v1alpha2/cri.go

```diff
@@ -248,65 +249,83 @@ func (c *CriManager) RunPodSandbox(ctx context.Context, r *runtime.RunPodSandbox
 		return nil, fmt.Errorf("failed to make sandbox pouch config for pod %q: %v", config.Metadata.Name, err)
 	}
 
-	sandboxName := makeSandboxName(config)
-
-	createResp, err := c.ContainerMgr.Create(ctx, sandboxName, createConfig)
+	// generate the id of sandbox meta and store it.
+	podSandboxID, err := c.generateSandboxID()
 	if err != nil {
-		return nil, fmt.Errorf("failed to create a sandbox for pod %q: %v", config.Metadata.Name, err)
+		return nil, err
 	}
-	id := createResp.ID
-
-	// once sandbox container created, we are obligated to store it.
 	sandboxMeta := &SandboxMeta{
-		ID:      id,
-		Config:  config,
-		Runtime: config.Annotations[anno.KubernetesRuntime],
+		ID: podSandboxID,
+	}
+	if err := c.SandboxStore.Put(sandboxMeta); err != nil {
+		return nil, err
 	}
 
+	// If running sandbox failed, clean up the container.
+	defer func() {
+		if !removeContainerErr {
+			// should not remove the sandbox container metadata from sandboxStore
+			// until it was removed by pouchd.
+			if err := c.SandboxStore.Remove(podSandboxID); err != nil {
+				logrus.Errorf("failed to remove the metadata of container %q from sandboxStore: %v", podSandboxID, err)
+			}
+		}
+	}()
+
 	if _, ok := config.Annotations[anno.LxcfsEnabled]; ok {
 		enableLxcfs, err := strconv.ParseBool(config.Annotations[anno.LxcfsEnabled])
 		if err != nil {
 			return nil, err
 		}
 		sandboxMeta.LxcfsEnabled = enableLxcfs
+
+		if err := c.SandboxStore.Put(sandboxMeta); err != nil {
+			return nil, err
+		}
+	}
+
+	sandboxName := makeSandboxName(config)
+
+	createResp, err := c.ContainerMgr.Create(ctx, sandboxName, createConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create a sandbox for pod %q: %v", config.Metadata.Name, err)
 	}
+	sandboxContainerID := createResp.ID
 
+	// once sandbox container created, we are obligated to store it.
+	sandboxMeta.SandboxContainerID = sandboxContainerID
+	sandboxMeta.Config = config
+	sandboxMeta.Runtime = config.Annotations[anno.KubernetesRuntime]
 	if err := c.SandboxStore.Put(sandboxMeta); err != nil {
 		return nil, err
 	}
 
 	// If running sandbox failed, clean up the container.
 	defer func() {
 		if retErr != nil {
-			if err := c.ContainerMgr.Remove(ctx, id, &apitypes.ContainerRemoveOptions{Volumes: true, Force: true}); err != nil {
-				logrus.Errorf("failed to remove container when running sandbox failed %q: %v", id, err)
-				return
-			}
-			// should not remove the sandbox container metadata from sandboxStore
-			// until it was removed by pouchd.
-			if err := c.SandboxStore.Remove(id); err != nil {
-				logrus.Errorf("failed to remove the metadata of container %q from sandboxStore: %v", id, err)
+			if err := c.ContainerMgr.Remove(ctx, sandboxContainerID, &apitypes.ContainerRemoveOptions{Volumes: true, Force: true}); err != nil {
```

Comment:
Can we add some test cases for this situation?

Reply:
Yeah， I have a test locally:
1. return err directly between success to create sandbox  and start
2. return err directly when remove container for pouchd
3. runpodsandbox
```
# crictl runp pod-config.yaml 
DEBU[0000] RunPodSandboxRequest: &RunPodSandboxRequest{Config:&PodSandboxConfig{Metadata:&PodSandboxMetadata{Name:test-remove-sandbox,Uid:hdishd83dddidwnduwk28bcsb,Namespace:default,Attempt:1,},Hostname:,LogDirectory:,DnsConfig:nil,PortMappings:[],Labels:map[string]string{},Annotations:map[string]string{},Linux:nil,},} 
DEBU[0000] RunPodSandboxResponse: nil                   
FATA[0000] run pod sandbox failed: rpc error: code = Unknown desc = test return error when run PodSandbox
```
4. pouch ps -a
```
# pouch ps -a
Name                                                              ID       Status         Created         Image                                                                 Runtime
k8s_POD_test-remove-sandbox_default_hdishd83dddidwnduwk28bcsb_1   03e587   created        2 minutes ago   registry.cn-hangzhou.aliyuncs.com/google-containers/pause-amd64:3.0   runc
```
5. pouch inspect
```
# pouch inspect 03e587
[
    ......          
            "Labels": {
                "io.kubernetes.pouch.type": "sandbox"
    ......      
]
```
6. ListPodSandbox
```
# crictl pods
DEBU[0000] ListPodSandboxRequest: &ListPodSandboxRequest{Filter:&PodSandboxFilter{Id:,State:nil,LabelSelector:map[string]string{},},} 
DEBU[0000] ListPodSandboxResponse: &ListPodSandboxResponse{Items:[&PodSandbox{Id:03e5874bc39ce08f70f260c00617f81d14242bf39cf53075022071d1af0a66bf,Metadata:&PodSandboxMetadata{Name:test-remove-sandbox,Uid:hdishd83dddidwnduwk28bcsb,Namespace:default,Attempt:1,},State:SANDBOX_NOTREADY,CreatedAt:1541681197387544250,Labels:map[string]string{},Annotations:map[string]string{},}],} 
POD ID              CREATED             STATE               NAME                  NAMESPACE           ATTEMPT
03e5874bc39ce       5 minutes ago       NotReady            test-remove-sandbox   default             1
```

Everything looks good. ^_^